### DISCUSSION TITLE
Multiple error strategies for the same process

### DISCUSSION THREAD

**Discussion by davidmasp** [nextflow-io/nextflow#2130] [Q&A] [answered]
I am trying to run a process with multiple exceptions for different error codes. Basically a process can die from memory for instance or from an internal error I want nextflow to ignore. Is there a nice way of doing that?

I have tried

```
errorStrategy { task.exitStatus == 124 ? 'retry' : {task.exitStatus == 123 ? 'ignore' : 'terminate'} }
```

but with this I get this error instead of ignoring it -> `Execution aborted due to an unexpected error `

I also tried this but seems it's not accepted by the syntax

```
  errorStrategy switch(task.exitStatus) {
    case 124:
      'retry'
    case 123:
      'ignore'
    default:
      'terminate'
  }
```

**Comment by abhi18av** [comment]
Hi @davidmasp ,

This is a really interesting problem and I wonder if we could try to use one of the following approaches 


1.  `nested ternary` might be worth exploring 

```groovy
errorStrategy { task.exitStatus == 124 ? 'retry' :  task.exitStatus == 123 ? 'ignore' : 'terminate' }

``` 
This idea is based on an analogues behavior in the JS world regarding [nested ternary operators](https://www.rithmschool.com/blog/using-the-javascript-ternary-operator) which make code somewhat "cleaner".

2. Use of a groovy function instead of a closure

```groovy

def myExitStrategy(exitStatus) {

switch(exitStatus) {
    case 124:
      'retry'
    case 123:
      'ignore'
    default:
      'terminate'
  }
}


process XYZ { 

...
errorStrategy myExitStrategy(task.exitStatus)

...
}
```


Curious to hear which one gets us closer to the goal.

Disclaimer: I haven't used either one in my own code so far (haven't come across a similar situation at all), so take these suggestions with a grain of salt.

**Reply by davidmasp** [reply]
Thanks for the input, I have to learn proper groovy eventually...

So the first option (1) works in my test set and it does actually look way more elegant than my if statement. The second I wasn't able to make it work could be I missed something.

**Reply by abhi18av** [reply]
Thanks for letting me know @davidmasp  👍

**Comment by davidmasp** [comment, marked as answer]
Okay, I managed to make it work using a if/else statement. Something like this:

```
  errorStrategy {
    if (task.exitStatus == 124){
      'retry'
    } else if (task.exitStatus == 123) {
      'ignore'
    } else {
      'terminate'
      }
    }
```


A full reproducible version of this should be available [here](https://github.com/davidmasp/hello/tree/multiple_errorStrategies)

EDIT:
As suggested by @abhi18av I have tested this solution which is more elegant and also works fine!

```
errorStrategy { task.exitStatus == 124 ? 'retry' :  (task.exitStatus == 123 ? 'ignore' : 'terminate') }
```

**Reply by pditommaso** [reply]
IMO this syntax is really ambiguous event to an experienced developer. The use of parenthesis is strongly suggested 


```
errorStrategy { task.exitStatus == 124 ? 'retry' :  (task.exitStatus == 123 ? 'ignore' : 'terminate') }
```
